torch_xla/csrc/aten_xla_type.cpp

```diff
@@ -866,12 +866,26 @@ AtenXlaType::convolution_backward_overrideable(
 
 at::Tensor& AtenXlaType::copy_(at::Tensor& self, const at::Tensor& src,
                                bool non_blocking) {
-  XLATensor self_tensor = bridge::GetXlaTensor(self);
+  c10::optional<XLATensor> self_tensor = bridge::TryGetXlaTensor(self);
   c10::optional<XLATensor> src_tensor = bridge::TryGetXlaTensor(src);
-  if (src_tensor) {
-    XLATensor::copy_(self_tensor, *src_tensor);
+
+  if (!src_tensor) {
+    TORCH_ASSERT(self_tensor);
+    self_tensor.SetTensor(CopyTensor(src, self_tensor->scalar_type()));
+  } else if (!self_tensor) {
+    // TODO: Is self_tensor good enough?  I don't think so... therefore
+    // the hack below:
+    //
+    // Do not mark the tensor creation as writeable to not discard the XLA tensor
```

Comment:
Comment no more valid.

Reply:
I can delete the comment. However, I am curious to know if I need to do all this faffing around in the first place. Ideally, I want to write the code like this:

```
  if (!src_tensor) {
    XLA_CHECK(self_tensor);
    self_tensor.SetTensor(CopyTensor(src, self_tensor->scalar_type()));
  } else if (!self_tensor) {
    at::Tensor t = CopyTensor(*self_tensor, self.scalar_type());
    const_cast<at::Tensor&>(self).unsafeGetTensorImpl()->shallow_copy_from(
        t.getIntrusivePtr());
```

i.e., not going through the XLA tensor list at all. Is that sound?

Reply:
> I can delete the comment. However, I am curious to know if I need to do all this faffing around in the first place. Ideally, I want to write the code like this:
> 
> ```
>   if (!src_tensor) {
>     XLA_CHECK(self_tensor);
>     self_tensor.SetTensor(CopyTensor(src, self_tensor->scalar_type()));
>   } else if (!self_tensor) {
>     at::Tensor t = CopyTensor(*self_tensor, self.scalar_type());
>     const_cast<at::Tensor&>(self).unsafeGetTensorImpl()->shallow_copy_from(
>         t.getIntrusivePtr());
> ```
> 
> i.e., not going through the XLA tensor list at all. Is that sound?

Wait, what? :grin:
You do need to get a CPU tensor our of the source (XLA) tensor.
How the above code seven work?
It seems to be copying self into self...

Reply:
I typoed!

```
  if (!src_tensor) {
    XLA_CHECK(self_tensor);
    self_tensor.SetTensor(CopyTensor(src, self_tensor->scalar_type()));
  } else if (!self_tensor) {
    at::Tensor t = CopyTensor(*src, self.scalar_type());
    const_cast<at::Tensor&>(self).unsafeGetTensorImpl()->shallow_copy_from(
        t.getIntrusivePtr());
```